### PR TITLE
Implemented the Moran process on graphs

### PR DESCRIPTION
**Pull request by marcharper**
[Slides here](http://www.nimbios.org/tutorials/gametheoryfiles/evolutionary_graph_theory_back_Broom.pdf) for context.

### CODE REVIEW

File: axelrod/__init__.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 # The order of imports matters!
 from .version import __version__
+from . import graph
```
Comment (drvinceknight) [inline axelrod/__init__.py]: I would have thought this line wasn't needed? Am I wrong?

File: axelrod/graph.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
+"""
+Weighted undirected sparse graphs.
+
+Original source:
+https://github.com/marcharper/stationary/blob/master/stationary/utils/graph.py
```
Comment (drvinceknight) [inline axelrod/graph.py]: Not against this. A couple of initial thoughts:

1. As you have built this for your stationary library you no doubt have a reason but why we wouldn't just use networkx (which would have the quick lookups we want here)? (1 negative side is adding another dependency to the library) 
2. Can we add some tests please?
3. The spatial tournament does not necessarily need a graph object (it just loops through the edges once via the match generator) but if we're going to have an `axelrod.graph` class we should also use that there (for consistency). I see two approaches:
     1. We allow the spatial tournament to take both a list of edges of one of these graph objects (ensures backwards compatibility);
     2. We let the Moran process take a list of edges and obfuscate the building of this underlying graph object there. I think this would be my preferred approach.

File: axelrod/graph.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    graph = Graph(directed=directed)
+    edges = []
+    for i in range(length - 1):
+        edges.append((i, i+1))
+    edges.append((length - 1, 0))
+    graph.add_edges(edges)
+    return graph
+
+
+def complete_graph(length, directed=False):
```
Comment (drvinceknight) [inline axelrod/graph.py]: Do we need these extra functions? Networkx can be used to easily get a variety of graphs. (Just a thought)

File: axelrod/graph.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def vertices(self):
+        """Returns the set of vertices of the graph."""
+        return list(self.out_mapping.keys())
+
+    def out_dict(self, source):
+        """Returns a dictionary of the outgoing edges of source with weights."""
+        return self.out_mapping[source]
+
+    def out_vertices(self, source):
```
Comment (drvinceknight) [inline axelrod/graph.py]: If undirected, why not just have a `adjacent_vertices` method? (To replace both the `in` and `out` ones?).

File: axelrod/graph.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/graph.py]: To be clear: If we go with my suggestion of `3.i` then that would be work for a different PR.

File: docs/tutorials/getting_started/moran.rst [hunk 5]
```diff
@@ -70,6 +70,50 @@ function like :code:`takewhile` from :code:`itertools`):
...
     ...         break
     >>> mp.population_distribution()
-    Counter({'Tit For Tat': 4})
+    Counter({'Cooperator': 4})
+
+
+Moran Process on Graphs
+-----------------------
+
+The library also provides a graph-based Moran process with `MoranProcessGraph`.
```
Comment (drvinceknight) [inline docs/tutorials/getting_started/moran.rst]: Would you be able to include a reference text/paper in the bibliography?

File: axelrod/graph.py (hunk 2 above)
Comment (marcharper) [inline axelrod/graph.py]: I figured this would come up which is why I didn't add tests yet. We can switch this out for another library, I just used what I had laying around. It's a bit overkill since we don't need the weights (yet?), but it optimized in the right ways -- sparse implementation, precomputed neighbors. So we could also just simplify it to parts needed (relatively easy).

I'm not opposed to networkX but I'm not familiar with its API, what internal representation it uses (adjacency matrix? sparse matrix? etc.).

I agree re: consistent API with spatial tournaments -- we could also use `isinstance` to be flexible on the input, and also have the `__repr__` method just output the list of edges. There are cases however that a list of edges doesn't cover, like some types of random graphs.

File: axelrod/graph.py (hunk 4 above)
Comment (marcharper) [inline axelrod/graph.py]: We may want to allow the graphs to be directed, see e.g. [here](http://www.usma.edu/nsc/siteassets/sitepages/publications/a%20novel%20analytical%20method%20for%20evolutionary%20graph.pdf).

File: axelrod/graph.py (hunk 3 above)
Comment (marcharper) [inline axelrod/graph.py]: I needed them for testing...

File: docs/tutorials/getting_started/moran.rst (hunk 5 above)
Comment (marcharper) [inline docs/tutorials/getting_started/moran.rst]: Yes I can add a reference.

File: axelrod/__init__.py (hunk 1 above)
Comment (drvinceknight) [inline axelrod/__init__.py]: > It depends on how you access the graph class in code, axelrod.graph.Graph will fail

I thought it wouldn't fail but happy as it is :)

File: axelrod/graph.py (hunk 4 above)
Comment (drvinceknight) [inline axelrod/graph.py]: Cool.

File: axelrod/graph.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/graph.py]: > There are cases however that a list of edges doesn't cover, like some types of random graphs.

That's the killer point right there! I guess that also includes the possibility of graphs that change over time etc which also rules out networkX (which really is a data representation library coupled with graph theoretic algorithms as opposed to the flexibility we might need here).

I suggest going for flexibility on the input (so we can open an issue for my suggestion of `3.i`).

File: axelrod/graph.py (hunk 3 above)
Comment (drvinceknight) [inline axelrod/graph.py]: Cool, we can also use these in some of the tests for Spatial Tournaments (not in this PR).

File: axelrod/graph.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/graph.py]: In order to get the `__repr__` method to output the list of edges I expect we could use an `edges` attribute or otherwise? That could be a useful attribute to have anyway?

Comment (marcharper) [inline axelrod/graph.py]: Yeah, we can save the input edges, or output a sorted list.

File: axelrod/graph.py (hunk 3 above)
Comment (marcharper) [inline axelrod/graph.py]: I started to add more, an N x M grid could be useful... we can just add them as we go.

File: axelrod/graph.py (hunk 2 above)
Comment (marcharper) [inline axelrod/graph.py]: Turns out we need the graph to be weighted for some applications in the literature. If the reproduction graph is weighted then it changes the death probabilities in the birth-death case.

**Comment by drvinceknight** [comment]
There's a py3.6 failure (on the merge with master).

File: docs/reference/bibliography.rst [hunk 6]
```diff
@@ -26,6 +26,7 @@ documentation.
 .. [PRISON1998] LIFL (1998) PRISON. Available at: http://www.lifl.fr/IPD/ipd.frame.html (Accessed: 19 September 2016).
 .. [Robson1989] Robson, Arthur, (1989), EFFICIENCY IN EVOLUTIONARY GAMES: DARWIN, NASH AND SECRET HANDSHAKE, Working Papers, Michigan - Center for Research on Economic & Social Theory, http://EconPapers.repec.org/RePEc:fth:michet:89-22.
 .. [Singer-Clark2014] Singer-Clark, T. (2014). Morality Metrics On Iterated Prisoner’s Dilemma Players.
+.. [Shakarian2013] Shakarian, P., Roos, P. & Moores, G. A Novel Analytical Method for Evolutionary Graph Theory Problems.
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: Thanks for this: could you cite it in the moran.rst file?

**Comment by drvinceknight** [comment]
> There's a py3.6 failure (on the merge with master).

@marcharper I took a look at the tests and pushed a fix. I pushed directly to this PR branch but feel free to change/modify/delete if you'd rather not.

> Thanks for this: could you cite it in the moran.rst file?

I also took the opportunity to pop the citation where I thought it should go. Only trying to be helpful so if you're not a fan please throw it out :)

**Comment by marcharper** [comment]
Thanks for the test fix!

File: axelrod/graph.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (meatballs) [inline axelrod/graph.py]: Might some of these methods be better as properties?

Comment (marcharper) [inline axelrod/graph.py]: Getters would make sense for e.g `vertices` not sure about the others. Maybe edges makes sense with a setter, essentially rewriting the graph. I'm open to suggestions.

Comment (meatballs) [inline axelrod/graph.py]: Let's leave it for now. If we fancy refactoring the graph, we can do so later.
